### PR TITLE
Add public token API to AppCheck

### PR DESCRIPTION
**Pull request by hsubox76**
Create public `getToken()` and `onTokenChanged()` methods.

Proposal (internal link):
go/app-check-sdk-get-token-3p-api

For docs reviewers:
The only public doc comments are in `packages/firebase/index.d.ts`

### CODE REVIEW

File: packages/app-check-types/index.d.ts [hunk 1]
```diff
@@ -36,6 +38,40 @@ export interface FirebaseAppCheck {
    * during `activate()`.
    */
   setTokenAutoRefreshEnabled(isTokenAutoRefreshEnabled: boolean): void;
+
+  /**
+   * Get the current App Check token. Attaches to the most recent
+   * in-flight request if one is present. Returns null if no token
+   * is present and no token requests are in-flight.
```
Comment (kevinthecheung) [inline packages/app-check-types/index.d.ts:45]:
```suggestion
   * is present and no token requests are in flight.
```

**Review by kevinthecheung — APPROVED** [review]

File: packages/app-check/src/api.ts [hunk 2]
```diff
@@ -82,3 +92,72 @@ export function setTokenAutoRefreshEnabled(
...
+   * It is added only for API consistency with the observer pattern, which
+   * we follow in JS APIs.
+   */
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  onCompletion?: () => void
+): Unsubscribe {
+  let nextFn: NextFn<AppCheckTokenResult> = () => {};
+  let errorFn: ErrorFn = () => {};
+  if ((onNextOrObserver as PartialObserver<AppCheckTokenResult>).next != null) {
+    nextFn = (onNextOrObserver as PartialObserver<AppCheckTokenResult>).next!;
```
Comment (Feiyang1) [inline packages/app-check/src/api.ts:150]: We need to bind the functions (next, error) to `onNextOrObserver`, otherwise we will trigger errors with observers created from rxjs.
see https://github.com/firebase/firebase-js-sdk/issues/4937

Comment (Feiyang1) [reply packages/app-check/src/api.ts:150]: Is the `!` necessary? Typscript should know it's not null inside the if block.

File: packages/app-check/src/internal-api.ts [hunk 3]
```diff
@@ -306,9 +326,13 @@ function notifyTokenListeners(
 
   for (const listener of listeners) {
     try {
-      listener(token);
+      listener.listener(token);
```
Comment (Feiyang1) [inline packages/app-check/src/internal-api.ts:329]: This feels weird with 2 `listener`s. Can you rename the field name to something different.  Perhaps we should just store observers in `state. tokenListeners` instead of `AppCheckTokenListenerInternal`s.

File: packages/app-check/src/api.test.ts [hunk 4]
```diff
@@ -86,4 +96,137 @@ describe('api', () => {
...
+      );
+
+      expect(getState(app).tokenListeners.length).to.equal(2);
+
+      await getToken(app, fakePlatformLoggingProvider);
+
+      expect(listener2).to.be.calledWith({
+        token: fakeRecaptchaAppCheckToken.token
+      });
+      expect(errorFn1).to.be.calledOnce;
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply packages/app-check/src/api.test.ts:174]: This should be fixed. The error callback will be called if getToken fails, with the error field from getToken.

File: packages/app-check-interop-types/index.d.ts [hunk 5]
```diff
@@ -24,10 +24,10 @@ export interface FirebaseAppCheckInternal {
   // registered at the same time for one or more FirebaseAppAttestation instances. The
   // listeners call back on the UI thread whenever the current token associated with this
   // FirebaseAppAttestation changes.
-  addTokenListener(listener: AppCheckTokenListener): void;
+  addTokenListener(listener: (token: AppCheckTokenResult) => void): void;
 
   // Unregisters a listener to changes in the token state.
-  removeTokenListener(listener: AppCheckTokenListener): void;
+  removeTokenListener(listener: (token: AppCheckTokenResult) => void): void;
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply packages/app-check-interop-types/index.d.ts:30]: Leftover from another approach, fixed.

File: packages/app-check/src/api.ts (hunk 2 above)
Comment (hsubox76) [reply packages/app-check/src/api.ts:150]: 1) Binding: done.
2) I don't understand why Typescript complains without the `!` but it does.

File: packages/app-check/src/internal-api.ts (hunk 3 above)
Comment (hsubox76) [reply packages/app-check/src/internal-api.ts:329]: I converted these to observers.

File: packages/app-check/src/internal-api.ts [hunk 6]
```diff
@@ -264,7 +277,11 @@ function createTokenRefresher(
 
       // getToken() always resolves. In case the result has an error field defined, it means the operation failed, and we should retry.
       if (result.error) {
-        throw result.error;
+        if (onError) {
+          onError(result.error);
```
Comment (Feiyang1) [inline packages/app-check/src/internal-api.ts:281]: This is redundant. Notifying listeners are already handled in `getToken()` which calls `notifyTokenListeners()`.

Here we should always throw to trigger retries. (Just revert the changes)

**Review by Feiyang1 — CHANGES_REQUESTED** [review]

File: packages/app-check/src/api.test.ts [hunk 7]
```diff
@@ -86,4 +97,171 @@ describe('api', () => {
...
+        app,
+        fakePlatformLoggingProvider,
+        true
+      );
+    });
+    it('getToken() throws errors returned with token', async () => {
+      const app = getFakeApp({ automaticDataCollectionEnabled: true });
+      const fakePlatformLoggingProvider = getFakePlatformLoggingProvider();
+      stub(internalApi, 'getToken').resolves({
+        token: 'a-token-string',
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply packages/app-check/src/api.test.ts:118]: Clarified.

File: packages/app-check/src/internal-api.ts (hunk 6 above)
Comment (hsubox76) [reply packages/app-check/src/internal-api.ts:281]: Reverted.

**Review by Feiyang1 — APPROVED** [review]
